### PR TITLE
Craig 291: Subnet form enhancements

### PR DESCRIPTION
**Pull request by jvallexm**

### CODE REVIEW

File: playground/src/App.js [hunk 1]
```diff
@@ -11,201 +11,95 @@ import {
...
+        disableSubnetSaveCallback={disableSubnetSaveCallback}
+        invalidCallback={invalidCallback}
+        invalidTextCallback={invalidTextCallback}
+        networkAcls={["example-acl-1", "example-acl-2"]}
+        enabledPublicGateways={[1, 2, 3]}
+        subnetListCallback={subnetListCallback}
+        propsMatchState={propsMatchState}
+        invalidSubnetCallback={() => {
+          return true;
+        }}
```
Comment (isaaceichelberger) [inline playground/src/App.js:92]: when this is set to true, i get an error when typing in the cidr field
```
this.props.componentDidUpdateCallback is not a function
TypeError: this.props.componentDidUpdateCallback is not a function
    at SubnetForm.componentDidUpdate (http://localhost:3001/static/js/bundle.js:6542:16)
    at commitLayoutEffectOnFiber (http://localhost:3001/static/js/bundle.js:127469:34)
    at commitLayoutMountEffects_complete (http://localhost:3001/static/js/bundle.js:128610:13)
    at commitLayoutEffects_begin (http://localhost:3001/static/js/bundle.js:128599:11)
    at commitLayoutEffects (http://localhost:3001/static/js/bundle.js:128545:7)
    at commitRootImpl (http://localhost:3001/static/js/bundle.js:130455:9)
    at commitRoot (http://localhost:3001/static/js/bundle.js:130335:9)
    at performSyncWorkOnRoot (http://localhost:3001/static/js/bundle.js:129844:7)
    at flushSyncCallbacks (http://localhost:3001/static/js/bundle.js:117847:26)
    at flushSyncCallbacksOnlyInLegacyMode (http://localhost:3001/static/js/bundle.js:117829:9)
```

File: src/components/forms/SubnetTileForm.js [hunk 2]
```diff
@@ -65,21 +65,51 @@ class SubnetTileForm extends React.Component {
...
-              shouldDisableGatewayToggle={this.shouldDisableGatewayToggle}
-            />
-          ))}
+          {subnetMap.map((subnet, index) => {
+            if (
+              this.props.advanced &&
+              !contains(this.props.select_zones, index + 1)
+            ) {
+              return (
+                <SubnetForm // change so doesn't show buttons
```
Comment (isaaceichelberger) [inline src/components/forms/SubnetTileForm.js:74]: is this comment a later thing or still a todo for this issue

File: src/components/forms/SubnetTierForm.js [hunk 3]
```diff
@@ -6,19 +6,24 @@ import {
...
 
 class SubnetTierForm extends React.Component {
   constructor(props) {
     super(props);
     this.state = { ...this.props.data };
+    if (!this.state.select_zones) {
+      let zones = buildNumberDropdownList(this.state.zones, 1);
+      this.state.select_zones = [];
+      zones.forEach((zone) => this.state.select_zones.push(Number(zone)));
+    }
```
Comment (isaaceichelberger) [inline src/components/forms/SubnetTierForm.js:26]: zones are displaying [1, 3, 2] - can we make them in order?

**Review by isaaceichelberger — CHANGES_REQUESTED** [review]
another issue i've noticed is if you change the name on the subnet tile to `""` it will disable the public gateways toggle and there is no way to turn it back on

File: src/components/forms/SubnetTileForm.js (hunk 2 above)
Comment (jvallexm) [reply src/components/forms/SubnetTileForm.js:74]: this is an old comment will delete

File: src/components/forms/SubnetTierForm.js (hunk 3 above)
Comment (jvallexm) [reply src/components/forms/SubnetTierForm.js:26]: i think it defaults to putting the selected ones first but i'll try

Comment (jvallexm) [reply src/components/forms/SubnetTierForm.js:26]: yeah this is not possible

File: src/components/forms/SubnetForm.js [hunk 4]
```diff
@@ -39,6 +39,9 @@ class SubnetForm extends React.Component {
   }
 
   render() {
+    function cidrIsValid(cidr) {
```
Comment (evpirnia) [inline src/components/forms/SubnetForm.js:42]: do we want this function inside the render? we could make it like the handleToggle function

File: src/components/forms/SubnetTierForm.js [hunk 5]
```diff
@@ -267,6 +330,10 @@ SubnetTierForm.propTypes = {
   enableModal: PropTypes.func,
```
Comment (evpirnia) [inline src/components/forms/SubnetTierForm.js:330]: there are two hides in this form - one in data and one in props. do we need both?

File: storybook/src/stories/Forms/Subnets/SubnetTierForm.stories.js [hunk 6]
```diff
@@ -127,6 +127,18 @@ export default {
       type: { required: true }, // required prop or not
       control: "none",
     },
+    invalidSubnetCallback: {
```
Comment (evpirnia) [inline storybook/src/stories/Forms/Subnets/SubnetTierForm.stories.js:130]: missing subnetListCallback

**Review by evpirnia — CHANGES_REQUESTED** [review]

**Review by isaaceichelberger — COMMENTED** [review]
still seeing the issue where if you backspace name to "" it disables the public gateway toggle and it cannot be reenabled, other than that lgtm

File: src/components/forms/SubnetTileForm.js [hunk 7]
```diff
@@ -65,21 +65,52 @@ class SubnetTileForm extends React.Component {
...
-              data={subnet}
-              onSave={this.props.onSave}
-              isModal={this.props.isModal || this.props.readOnly}
-              componentDidUpdateCallback={this.childSubnetHasChanged}
-              networkAcls={this.props.networkAcls}
-              disableSaveCallback={this.props.disableSaveCallback}
-              shouldDisableGatewayToggle={this.shouldDisableGatewayToggle}
-            />
-          ))}
+          {subnetMap.map((subnet, index) => {
```
Comment (evpirnia) [inline src/components/forms/SubnetTileForm.js:68]: does this have Alex's fix? it looks like subnets is still indented

Screenshot 2023-04-26 at 12 22 22 PM

File: src/components/forms/SubnetTierForm.js (hunk 5 above)
Comment (jvallexm) [reply src/components/forms/SubnetTierForm.js:330]: yes, needed for different use cases

File: src/components/forms/SubnetTileForm.js (hunk 7 above)
Comment (jvallexm) [reply src/components/forms/SubnetTileForm.js:68]: it should? i pulled main before i started but i'll check

Comment (jvallexm) [reply src/components/forms/SubnetTileForm.js:68]: yes it does, not sure what's up here

**Review by evpirnia — APPROVED** [review]

**Review by isaaceichelberger — APPROVED** [review]

**Pull request merged**